src/Aspire.Dashboard/Model/DashboardClient.cs

```diff
@@ -0,0 +1,309 @@
+                                    if (_resourceByName.Remove(change.Delete.ResourceName, out var removed))
+                                    {
+                                        changes.Add(new(ResourceViewModelChangeType.Delete, removed));
+                                    }
+                                    else
+                                    {
+                                        Debug.Fail("Attempt to remove an unknown resource view model.");
+                                    }
+                                }
+                            }
+                        }
+                        else
+                        {
+                            throw new FormatException("Unsupported response kind: " + response.KindCase);
+                        }
+                    }
+
+                    if (changes is not null)
+                    {
+                        foreach (var channel in _outgoingChannels)
+                        {
+                            // TODO send batches over the channel instead of individual items? They are batched downstream however
+                            foreach (var change in changes)
+                            {
+                                await channel.Writer.WriteAsync(change, cancellationToken).ConfigureAwait(false);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        static Uri GetAddressUri(string variableName, string defaultValue)
+        {
+            try
+            {
+                var uri = Environment.GetEnvironmentVariable(variableName) ?? defaultValue;
+
+                return new Uri(uri);
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidOperationException($"Error parsing URIs from environment variable '{variableName}'.", ex);
+            }
+        }
+    }
+
+    public string ApplicationName => _applicationName ?? "";
+
+    ResourceViewModelSubscription IDashboardClient.SubscribeResources()
+    {
+        EnsureInitialized();
+
+        lock (_lock)
+        {
+            // There are two types of channel in this class. This is not a gRPC channel.
+            // It's a producer-consumer queue channel, used to push updates to subscribers
+            // without blocking the producer here.
+            var channel = Channel.CreateUnbounded<ResourceViewModelChange>(
+                new UnboundedChannelOptions { AllowSynchronousContinuations = false, SingleReader = true, SingleWriter = true });
+
+            ImmutableInterlocked.Update(ref _outgoingChannels, static (set, channel) => set.Add(channel), channel);
+
+            return new ResourceViewModelSubscription(
+                InitialState: _resourceByName.Values.ToImmutableArray(),
+                Subscription: StreamUpdates());
+
+            async IAsyncEnumerable<ResourceViewModelChange> StreamUpdates()
```

Comment:
FYI this has the same problem as I commented on here: https://github.com/dotnet/aspire/pull/1348#discussion_r1428707324

Subscriptions won't be stopped when they're canceled.

Reply:
I'm going to take a separate pass through all this and ensure nothing's running longer or more frequently than it should. I saw your comment there and will address it in a subsequent PR as the problem already exists on main and this PR is large enough as is.